### PR TITLE
Fix Alpine build

### PR DESCRIPTION
**Pull request by akoeplinger**
See commit messages for details.

Closes #3212.

### CODE REVIEW

File: mono/mini/exceptions-arm.c [hunk 1]
```diff
@@ -14,9 +14,9 @@
 #include <string.h>
 
 #ifndef MONO_CROSS_COMPILE
-#ifdef HAVE_ASM_SIGCONTEXT_H
+#ifdef PLATFORM_ANDROID
 #include <asm/sigcontext.h>
-#endif  /* def HAVE_ASM_SIGCONTEXT_H */
+#endif  /* def PLATFORM_ANDROID */
 #endif
```
Comment (migueldeicaza) [inline mono/mini/exceptions-arm.c]: This change looks strange, it would now only include this file if running on Android.

Why was this necessary?

We have two options, if we want it included on Android or if the header file is present, perhaps we should change that to be HAVE_ASM_SIGCONTEXT_H || PLATFORM_ANDROID

If the purpose was to avoid including the header file, I would like to know why :-)

File: mono/io-layer/processes.h [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #ifndef _WAPI_PROCESSES_H_
 #define _WAPI_PROCESSES_H_
 
+#include <sys/types.h>
```
Comment (migueldeicaza) [inline mono/io-layer/processes.h]: We should change this to:

```
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
```

File: libgc/os_dep.c [hunk 3]
```diff
@@ -26,7 +26,7 @@
 #     define __KERNEL__
 #     include <asm/signal.h>
 #     undef __KERNEL__
-#   else
+#   elif defined(__GLIBC__)
```
Comment (migueldeicaza) [inline libgc/os_dep.c]: For Alpine, perhaps we can detect and disable Boehm?

I am not crazy about adding libc-specific defines.

Comment (akoeplinger) [inline libgc/os_dep.c]: If you look a few lines down there are already `__GLIB__` defines, the purpose is to skip those completely as logic like `#if 2 <= __GLIBC__` doesn't work if the symbol isn't defined at all.

File: mono/io-layer/processes.h (hunk 2 above)
Comment (akoeplinger) [inline mono/io-layer/processes.h]: We do not guard sys/types.h anywhere else in io-layer: https://github.com/mono/mono/search?p=2&q=%22sys%2Ftypes.h%22+path%3Amono%2Fio-layer&type=Code&utf8=%E2%9C%93

File: mono/mini/exceptions-arm.c (hunk 1 above)
Comment (akoeplinger) [inline mono/mini/exceptions-arm.c]: Yes, the purpose is to avoid including this header file, Alpine sets HAVE_ASM_SIGCONTEXT_H and including it causes struct redefinition errors as the types are already defined in signal.h.

From the commit message:
```
[mini] Fix asm/sigcontext.h include in exceptions-arm.h

We include the header under the PLATFORM_ANDROID ifdef in mono-sigcontext.h as well [1].
It was probably added because older Bionic didn't include this from signal.h [2].

Fixes the build on Alpine/musl-libc.

[1] https://github.com/mono/mono/blob/19a99abf1680c85f8ff58c2b44ea8c15bb557ca5/mono/utils/mono-sigcontext.h#L5
[2] https://bugzilla.mozilla.org/show_bug.cgi?id=859580
```

**Comment by migueldeicaza** [comment]
Thanks for the comments Alexander.

Let us get this merged.
